### PR TITLE
feat: devops workflow setting refactor

### PR DESCRIPTION
**Pull request by laojun**
## I have checked the following points:
- [ ] I18n is finished and updated by cli
- [ ] Form fields validation is added and length is limited
- [ ] Display normally on small screen
- [ ] Display normally when some data is empty or null
- [ ] Display normally in english mode

## Which issue(s) this PR fixes:
Fixes #

- Fixes #your-issue_number
- [Erda Cloud Issue Link](paste your link here)

## Does this PR introduce a user interface change?

✅ Yes(screenshot is required)
image
image

## ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |    feat: devops workflow setting refactor          |
| 🇨🇳 中文    |      feat: 研发工作流设置重构       |

## Need cherry-pick to release versions?
❎ No

### CODE REVIEW

File: shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx [hunk 1]
```diff
@@ -0,0 +1,818 @@
...
+import { Button, Switch, Input, Select, Popconfirm, message, Tooltip, Spin } from 'antd';
+import { uuid } from 'common/utils';
+import { throttle, compact } from 'lodash';
+import { FlowType } from 'project/common/config';
+import ResizeObserver from 'rc-resize-observer';
+import { branchNameValidator, branchNameWithoutWildcard } from 'project/common/config';
+import { WithAuth } from 'user/common';
+import themeColor from 'app/theme-color.mjs';
+import i18n from 'i18n';
+interface IProps {
```
Comment (McDaddy) [inline shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx:27]: Leave an empty line between import and code

File: shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx [hunk 2]
```diff
@@ -0,0 +1,818 @@
...
+  d.map((item) => {
+    return {
+      ...item,
+      id: uuid(),
+      openTempMerge: !!item?.policy?.tempBranch,
+      policy: item.policy && {
+        ...item.policy,
+        targetBranch: item.policy.targetBranch && {
+          ...item.policy.targetBranch,
+          cherryPick: item.policy.targetBranch.cherryPick || undefined,
```
Comment (McDaddy) [inline shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx:47]: is `|| undefined` required

File: shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx [hunk 3]
```diff
@@ -0,0 +1,818 @@
...
+              setUseData((prev) => [emptyData, ...prev]);
+              setEditData({ ...emptyData });
+            }}
+          >
+            {i18n.t('Add')}
+          </Button>
+        </WithAuth>
+      </div>
+      {useData.map((item) => {
+        const deleteCurData = () => {
```
Comment (McDaddy) [inline shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx:132]: maybe move these two func out of the map iteration is better.

File: shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx [hunk 4]
```diff
@@ -0,0 +1,818 @@
...
+          }
+        />
+        <span className="px-1 text-base">/</span>
+        <Select
+          bordered={false}
+          className="bg-default-06 w-[75px] h-[30px] rounded"
+          value={type}
+          onChange={(v) => onChange({ type: v, value: v === FlowType.MULTI_BRANCH ? `${val}/*` : val })}
+        >
+          <Select.Option value={FlowType.SINGLE_BRANCH}>无</Select.Option>
```
Comment (McDaddy) [inline shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx:502]: i18n

File: shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx [hunk 5]
```diff
@@ -0,0 +1,818 @@
...
+          autoFocus
+          placeholder={i18n.t('start with letters and can contain characters that are not wildcard')}
+          className="bg-default-06 border-none rounded"
+          value={data?.policy?.sourceBranch}
+          onChange={(e) => setData({ ...data, policy: { ...data.policy, sourceBranch: e.target.value } })}
+        />
+      ) : (
+        <Ellipsis
+          className="text-default-8 bg-default-06 rounded px-[11px] py-1 hover:bg-default-1 cursor-pointer h-[30px]"
+          title={data?.policy?.sourceBranch}
```
Comment (McDaddy) [inline shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx:476]: possible to show undefined?

File: shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx [hunk 6]
```diff
@@ -0,0 +1,818 @@
...
+              />
+            )}
+          </If>
+        </div>
+      </If>
+    </div>
+  );
+};
+
+const TargetBranchCard = (props: CardProps) => {
```
Comment (McDaddy) [inline shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx:594]: this file is too large and complex... better split into separate file

File: shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx (hunk 2 above)
Comment (laojun) [reply shell/app/modules/project/pages/settings/components/workflow-setting/branch-policy.tsx:47]: yes, cherrypick maybe "",  initial value needs to split(','),  ''.split(',') = [''], the expected value is []。
after initial  cherrypick could be ''

File: shell/app/modules/project/pages/settings/components/workflow-setting/devops-workflow.tsx [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+                });
+              },
+            },
+          ];
+        },
+      }
+    : undefined;
+
+  return (
+    <div className="">
```
Comment (McDaddy) [inline shell/app/modules/project/pages/settings/components/workflow-setting/devops-workflow.tsx:182]: empty class？

File: shell/app/modules/project/services/project-workflow.ts [hunk 8]
```diff
@@ -177,3 +133,37 @@ export const deleteFlowNode = apiCreator<(payload: { mergeID: number }) => DevFl
 export const getPipelineDetail = apiCreator<(payload: { pipelineID: number }) => BUILD.IPipelineDetail>(
   apis.getPipelineDetail,
 );
+
+export interface BranchPolicy {
```
Comment (McDaddy) [inline shell/app/modules/project/services/project-workflow.ts:137]: should defined in .d.ts?

**Review by McDaddy — APPROVED** [review]

**Pull request merged**